git/odb/tree.go

```diff
@@ -0,0 +1,123 @@
+package odb
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"syscall"
+)
+
+// Tree encapsulates a Git tree object.
+type Tree struct {
+	// Entries is the list of entries held by this tree.
+	Entries []*TreeEntry
+}
+
+var _ Object = (*Tree)(nil)
+
+// Type implements Object.ObjectType by returning the correct object type for
+// Trees, TreeObjectType.
+func (t *Tree) Type() ObjectType { return TreeObjectType }
+
+// Decode implements Object.Decode and decodes the uncompressed tree being
+// read. It returns the number of uncompressed bytes being consumed off of the
+// stream, which should be strictly equal to the size given.
+//
+// If any error was encountered along the way, that will be returned, along with
+// the number of bytes read up to that point.
+func (t *Tree) Decode(from io.Reader, size int64) (n int, err error) {
+	buf := bufio.NewReader(from)
+
+	var entries []*TreeEntry
+	for {
+		modes, err := buf.ReadString(' ')
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return n, err
+		}
+		n += len(modes)
+		modes = strings.TrimSuffix(modes, " ")
+		modes = fmt.Sprintf("%06s", modes)
```

Comment:
Good catch, this is a misconception on my part. Since Git removes leading zeros on the left, I assumed the wrong byte order and thought that adding those zeroes back would be significant (this is what "%06s" does, add `len(s)-6` `0`'s to the front of `%s`).

My byte-order assumption was wrong, so I removed this line in 39d000c7.